dataset.go

```diff
@@ -86,6 +114,70 @@ func flatElementsIterator(elems []*Element, elemChan chan<- *Element) {
 	}
 }
 
+// FlatDatasetIterator represents a simple iterator over the elements in a
+// Dataset.
+type FlatDatasetIterator interface {
+	// HasNext indicates if the iterator as another element.
+	HasNext() bool
+	// Next gets and returns the next element in the iterator.
+	Next() *Element
```

Comment:
This is backwards from the convention I've normally seen for iterators.

Generally, ``Next()`` returns the bool and some other method like ``Current()`` fetches the item.

Here is an example from the sql package's ``Row`` type, which uses ``Next()`` / ``Scan()``: https://golang.org/pkg/database/sql/#Rows

The MongoDB cursor type also uses ``Next()`` / ``Decode()`` in this way:

https://pkg.go.dev/go.mongodb.org/mongo-driver/mongo#Cursor

Reply:
I have seen patterns like this, but personally find them a little odd. `Next()` returning a bool seems odd to me, whereas `HasNext` makes that very clear and readable imo. I could rename `Next()` to be `NextValue()` to be clearer, but I'm not so sure that's really needed (as the rest of the fn signature makes that clear).